extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -14,43 +14,415 @@
+import {tryParseJson} from '../../../src/json';
 
-/** @const {string} */
+/** @const */
 const EXPERIMENT = 'amp-autocomplete';
 
-/** @const {string} */
+/** @const */
 const TAG = 'amp-autocomplete';
 
+/**
+ * Different filtering options.
+ * @enum {string}
+ */
+export const FilterType = {
+  SUBSTRING: 'substring',
+  PREFIX: 'prefix',
+  TOKEN_PREFIX: 'token-prefix',
+  FUZZY: 'fuzzy',
+  CUSTOM: 'custom',
+  NONE: 'none',
+};
+
 export class AmpAutocomplete extends AMP.BaseElement {
 
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
 
-    /** @private {string} */
-    this.myText_ = 'hello world';
+    /**
+     * The data extracted from the <script> tag optionally provided
+     * as a child. For use with static data.
+     * @private {?Array}
+     */
+    this.inlineData_ = null;
+
+    /**
+     * The reference to the <input> tag provided as a child.
+     * @private {?HTMLElement}
+     */
+    this.inputElement_ = null;
+
+    /**
+     * The value of the "filter" attribute on <autocomplete>.
+     * @private {string}
+     */
+    this.filter_ = '';
+
+    /**
+     * The value of the "min-characters" attribute on <autocomplete>.
+     * @private {number}
+     */
+    this.minChars_ = 1;
+
+    /**
+     * The value of the "max-entries" attribute on <autocomplete>.
+     * @private {?number}
+     */
+    this.maxEntries_ = null;
 
-    /** @private {?Element} */
+    /**
+     * The index of the active suggested item.
+     * @private {number}
+     */
+    this.activeIndex_ = -1;
+
+    /**
+     * The reference to the <div> of the active suggested item.
+     * @private {?Element}
+     */
+    this.activeElement_ = null;
+
+    /**
+     * The reference to the <div> that contains template-rendered children.
+     * @private {?Element}
+     */
     this.container_ = null;
   }
 
   /** @override */
   buildCallback() {
     userAssert(isExperimentOn(this.win, 'amp-autocomplete'),
         `Experiment ${EXPERIMENT} is not turned on.`);
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
-    this.element.appendChild(this.container_);
-    this.applyFillContent(this.container_, /* replacedContent */ true);
+
+    this.inlineData_ = this.getInlineData_();
+
+    const inputElements = childElementsByTag(this.element, 'INPUT');
+    userAssert(inputElements.length === 1,
+        `${TAG} should contain exactly one <input> child`);
+    this.inputElement_ = inputElements[0];
+
+    this.filter_ = userAssert(this.element.getAttribute('filter'),
+        `${TAG} requires "filter" attribute.`);
+    userAssert(isEnumValue(FilterType, this.filter_),
+        `Unexpected filter: ${this.filter_}`);
+
+    this.minChars_ = this.element.hasAttribute('min-characters') ?
+      parseInt(this.element.getAttribute('min-characters'), 10) : 1;
+    this.maxEntries_ = parseInt(this.element.getAttribute('max-entries'), 10);
```

Comment:
This can make `maxEntries` `NaN` (since `parseInt(null, 10)` would be `NaN`).

`NaN` is still falsy and works in most cases but to adhere to type definition of null or number, let's handle this like `minChar` above and check existence.